dev/compare_models.py

```diff
@@ -0,0 +1,307 @@
+        combinations = (dict(zip(param_dict.keys(), values))
+                        for values in product(*param_dict.values()))
+
+        for combination in combinations:
+
+            new_config = copy.deepcopy(initial_config)
+
+            for param in combination:
+
+                value = combination[param]
+                new_config[param] = value
+                new_config["log_directory"] = new_config["log_directory"] + \
+                    "-" + param + "=" + str(value)
+
+            config_list.append(copy.deepcopy(new_config))
+    # Change a single parameter for each test
+    else:
+        for param in param_dict:
+
+            new_config = copy.deepcopy(initial_config)
+
+            for value in param_dict[param]:
+
+                new_config[param] = value
+                new_config["log_directory"] = initial_config["log_directory"] + \
+                    "-" + param + "=" + str(value)
+                config_list.append(copy.deepcopy(new_config))
+
+    # CUDA problem when forking process
+    # https://github.com/pytorch/pytorch/issues/2517
+    mp.set_start_method('spawn')
+
+    # Run all configs on a separate process, with a maximum of n_gpus  processes at a given time
+    pool = mp.Pool(processes=len(initial_config["gpu"]))
+    if(args.n_iterations is not None):
+        n_iterations = args.n_iterations
+    else:
+        n_iterations = 1
+
+    results_df = pd.DataFrame()
+    for i in range(n_iterations):
+        if not args.fixed_split:
+            # Set seed for iteration
+            seed = random.randint(1, 10001)
+            for config in config_list:
+                config["random_seed"] = seed
+
+        validation_scores = pool.map(train_worker, config_list)
+        val_df = pd.DataFrame(validation_scores, columns=[
+            'log_directory', 'best_training_dice', 'best_training_loss', 'best_validation_dice', 'best_validation_loss'])
+        #results_df = pd.concat([results_df, temp_df])
+
+        if(args.run_test):
+            for config in config_list:
+                # Delete path_pred
+                path_pred = os.path.join(config['log_directory'], 'pred_masks')
+                if os.path.isdir(path_pred) and n_iterations > 1:
+                    try:
+                        shutil.rmtree(path_pred)
+                    except OSError as e:
+                        print("Error: %s - %s." % (e.filename, e.strerror))
+
+            test_scores = pool.map(test_worker, config_list)
+            test_df = pd.DataFrame(test_scores, columns=[
+                                   'log_directory', 'test_dice'])
+            combined_df = val_df.set_index('log_directory').join(
+                test_df.set_index('log_directory'))
+            combined_df = combined_df.reset_index()
+
+        else:
+            combined_df = val_df
+
+        results_df = pd.concat([results_df, combined_df])
+
+    # Compute avg, std, p-values
+    if(n_iterations > 1):
```

Comment:
do you want to separate statistics calculation in another script? e.g. i can imaging that running all the combinations will take a lot of time, so in case you want to play with the statistics (e.g. add another test, try other threshold values, etc.) you would not have to re-run all the training but you would just input your pandas DF in your compute_statistics module. just an idea.

Reply:
Yep good idea, will look into it